test/integration/master/secrets_enveloping_test.go

```diff
@@ -0,0 +1,302 @@
+	}
+}
+
+func runEnvelopeTest(t *testing.T, unSealSecretFunc unSealSecret, transformerConfigYAML, expectedEnvelopePrefix string) {
+	transformerConfig := parseTransformerConfigOrDie(t, transformerConfigYAML)
+
+	storageConfig := framework.SharedEtcd()
+	kubeAPIServer, err := startKubeApiWithEncryption(t, storageConfig, transformerConfigYAML)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+	defer kubeAPIServer.TearDownFn()
+
+	client, err := kubernetes.NewForConfig(kubeAPIServer.ClientConfig)
+	if err != nil {
+		t.Fatalf("error while creating client: %v", err)
+	}
+
+	ns, err := createTestNamespace(client, testNamespace)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+	defer func() {
+		client.CoreV1().Namespaces().Delete(ns.Name, metav1.NewDeleteOptions(0))
+	}()
+
+	_, err = createTestSecret(client, testSecret, ns.Name)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+
+	etcdPath := getETCDPath(storageConfig.Prefix)
+	response, err := readRawRecordFromETCD(kubeAPIServer, etcdPath)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+
+	if !bytes.HasPrefix(response.Kvs[0].Value, []byte(expectedEnvelopePrefix)) {
+		t.Errorf("expected secret to be enveloped by %s, but got %s",
+			expectedEnvelopePrefix, response.Kvs[0].Value)
+		return
+	}
+
+	// etcd path of the key is used as authenticated context - need to pass it to decrypt
+	ctx := value.DefaultContext([]byte(etcdPath))
+	// Envelope header precedes the payload
+	sealedData := response.Kvs[0].Value[len(expectedEnvelopePrefix):]
+	v, err := unSealSecretFunc(sealedData, ctx, transformerConfig)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+	if !strings.Contains(string(v), secretVal) {
+		t.Errorf("expected %q after decryption, but got %q", secretVal, string(v))
+	}
+
+	// Secrets should be un-enveloped on direct reads from Kube API Server.
+	s, err := client.CoreV1().Secrets(testNamespace).Get(testSecret, metav1.GetOptions{})
+	if secretVal != string(s.Data[secretKey]) {
+		t.Errorf("expected %s from KubeAPI, but got %s", secretVal, string(s.Data[secretKey]))
+	}
+}
+
+func startKubeApiWithEncryption(t *testing.T, storageConfig *storagebackend.Config,
+	transformerConfig string) (*kubeapiservertesting.TestServer, error) {
+	tempDir, err := ioutil.TempDir("", "secrets-encryption-test")
+	if err != nil {
+		return nil, fmt.Errorf("failed to create temp directory: %v", err)
+	}
+	defer os.RemoveAll(tempDir)
+
+	encryptionConfig, err := ioutil.TempFile(tempDir, "encryption-config")
+	if err != nil {
+		return nil, fmt.Errorf("error while creating temp file for encryption config %v", err)
+	}
+
+	if _, err := encryptionConfig.Write([]byte(transformerConfig)); err != nil {
+		return nil, fmt.Errorf("error while writing encryption config: %v", err)
+	}
+
+	kubeAPIOptions := []string{"--experimental-encryption-provider-config", encryptionConfig.Name()}
+	server, err := kubeapiservertesting.StartTestServer(t, kubeAPIOptions, storageConfig)
```

Comment:
The code could be simplified if we would use `StartTestServerOrDie()`.

Reply:
It feels that by calling StartTestServerOrDie, I would need to refactor startKubeApiWithEncryption back to startKubeApiWithEncryptionOrDie since it would no longer be possible to return an error in all cases.

Reply:
Fair enough. Let's leave it as-is then.